File: oci-start-mac/OciStart/Core/AuthService.swift
import Foundation

final class AuthService {
    private let client: APIClient

    init(client: APIClient = .shared) {
        self.client = client
    }

    func login(
        baseURL: String,
        username: String,
        password: String,
        verificationCode: String? = nil,
        mfaCode: String? = nil,
        rememberMe: Bool = true
    ) async throws {
        let config = try await client.fetchLoginPageMeta(baseURL: baseURL)
        if config.turnstileEnabled {
            throw APIError.serverMessage("服务器已启用人机验证，请在浏览器中登录。")
        }
        let encrypted: String
        if let publicKey = config.publicKey {
            guard let cipher = RSAHelper.encrypt(plainText: password, publicKeyBase64: publicKey) else {
                throw APIError.serverMessage("登录公钥无效，请重试。")
            }
            encrypted = cipher
        } else if config.usesLegacyHTML {
            encrypted = password
        } else {
            throw APIError.invalidResponse
        }

        let result = try await client.performLogin(
            baseURL: baseURL,
            username: username,
            password: encrypted,
            verificationCode: verificationCode,
            mfaCode: mfaCode,
            rememberMe: rememberMe
        )

        if result.success == false {
            throw APIError.serverMessage(result.message ?? "登录失败")
        }
    }

    func registerFirstUser(baseURL: String, username: String, password: String) async throws {
        try await client.registerFirstUser(baseURL: baseURL, username: username, password: password)
    }

    func sendVerificationCode(baseURL: String, username: String) async throws {
        try await client.sendLoginVerificationCode(baseURL: baseURL, username: username)
    }

    func loginPageMeta(baseURL: String) async throws -> APIClient.LoginPageMeta {
        try await client.fetchLoginPageMeta(baseURL: baseURL)
    }

    static func extractPublicKeyPublic(from html: String) -> String? {
        extractPublicKey(from: html)
    }

    func logout(baseURL: String) async {
        await client.performLogout(baseURL: baseURL)
    }

    func validateSession(baseURL: String) async -> Bool {
        await client.validateSession(baseURL: baseURL)
    }

    func loginFactorConfig(baseURL: String) async -> APIClient.LoginFactorConfig {
        await client.fetchLoginFactorConfig(baseURL: baseURL)
    }

    private static func extractPublicKey(from html: String) -> String? {
        let pattern = #"window\.RSA_PUBLIC_KEY\s*=\s*"([A-Za-z0-9+/=\r\n]+)""#
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: html, range: NSRange(html.startIndex..., in: html)),
              let range = Range(match.range(at: 1), in: html) else {
            return nil
        }
        return String(html[range])
    }
}

File: oci-start-win/src/OciStart/Core/AuthService.cs
using System.Net;
using System.Text.Json;

namespace OciStart.Core;

/// <summary>Login / logout / session probe — align Mac AuthService.</summary>
public sealed class AuthService
{
    private readonly ApiClient _api = ApiClient.Shared;

    public sealed class LoginResult
    {
        public bool? Success { get; set; }
        public string? RedirectUrl { get; set; }
        public string? Message { get; set; }
    }

    public async Task LoginAsync(
        string username,
        string password,
        string? verificationCode = null,
        string? mfaCode = null,
        bool rememberMe = true,
        CancellationToken ct = default)
    {
        var config = await _api.GetLoginBootstrapAsync(ct).ConfigureAwait(false);
        if (config.TurnstileEnabled)
            throw ApiError.Server("服务器已启用人机验证，请在浏览器中登录。");
        string encrypted;
        if (config.PublicKey != null)
            encrypted = RsaHelper.Encrypt(password, config.PublicKey)
                ?? throw ApiError.Server("登录公钥无效，请重试。");
        else if (config.UsesLegacyHtml)
            encrypted = password;
        else
            throw ApiError.InvalidResponse();

        var fields = new Dictionary<string, string>
        {
            ["username"] = username,
            ["password"] = encrypted,
            ["remember-me"] = rememberMe ? "true" : "false"
        };
        if (!string.IsNullOrWhiteSpace(verificationCode))
            fields["verificationCode"] = verificationCode!;
        if (!string.IsNullOrWhiteSpace(mfaCode))
            fields["mfaCode"] = mfaCode!;

        var (data, status) = await _api.PostFormAsync("/perform_login", fields, ct).ConfigureAwait(false);
        if (status == HttpStatusCode.Unauthorized)
        {
            var partial = JsonUtil.Deserialize<LoginResult>(data);
            throw ApiError.Server(partial?.Message ?? "用户名或密码错误");
        }

        if ((int)status is < 200 or >= 300)
            throw ApiError.Server($"登录失败 HTTP {(int)status}");

        if (data.Length == 0) return;

        var result = JsonUtil.Deserialize<LoginResult>(data);
        if (result?.Success == false)
            throw ApiError.Server(result.Message ?? "登录失败");
    }

    public async Task LogoutAsync(CancellationToken ct = default)
    {
        try
        {
            await _api.GetHtmlAsync("/logout", ct).ConfigureAwait(false);
        }
        catch
        {
            // best-effort
        }
        finally
        {
            _api.ClearCookies();
        }
    }

    public async Task<bool> ValidateSessionAsync(CancellationToken ct = default)
    {
        try
        {
            var data = await _api.GetJsonAsync("/api/userInfo", ct: ct).ConfigureAwait(false);
            var body = JsonUtil.Obj(data);
            if (body == null || !body.TryGetValue("success", out var success)
                || success.ValueKind != JsonValueKind.True || JsonUtil.Int(body, "code") != 200
                || !body.TryGetValue("data", out var user) || user.ValueKind != JsonValueKind.Object)
                return false;
            return user.TryGetProperty("username", out var username)
                && username.ValueKind == JsonValueKind.String && !string.IsNullOrWhiteSpace(username.GetString());
        }
        catch
        {
            return false;
        }
    }
}
